File: server-rs/src/routes/fs.rs
use axum::{Json, Router, extract::Query, routing::get};
use serde::Deserialize;
use serde_json::{Value, json};

use crate::AppState;
use crate::errors::ApiError;
use super::shared::home_dir;

pub(crate) fn fs_routes() -> Router<AppState> {
    Router::new().route("/api/fs/list", get(fs_list))
}

#[derive(Debug, Deserialize)]
struct FsListQuery {
    path: Option<String>,
}

async fn fs_list(Query(query): Query<FsListQuery>) -> Result<Json<Value>, ApiError> {
    let base = if let Some(ref p) = query.path {
        let p = p.trim();
        if p.is_empty() {
            home_dir()
        } else {
            std::path::PathBuf::from(p)
        }
    } else {
        home_dir()
    };

    let canonical = base
        .canonicalize()
        .map_err(|_| ApiError::bad_request("Cannot resolve path."))?;

    if !canonical.is_dir() {
        return Err(ApiError::bad_request("Path is not a directory."));
    }

    let mut dirs: Vec<Value> = Vec::new();
    let mut read_dir = std::fs::read_dir(&canonical)
        .map_err(|e| ApiError::bad_request(format!("Cannot read directory: {}", e)))?;

    while let Some(Ok(entry)) = read_dir.next() {
        let file_type = match entry.file_type() {
            Ok(ft) => ft,
            Err(_) => continue,
        };
        if !file_type.is_dir() {
            continue;
        }
        let name = entry.file_name().to_string_lossy().to_string();
        if name.starts_with('.') {
            continue;
        }
        let full_path = entry.path().to_string_lossy().to_string();
        let is_git = entry.path().join(".git").exists();
        dirs.push(json!({ "name": name, "path": full_path, "isGit": is_git }));
    }

    dirs.sort_by(|a, b| {
        let an = a["name"].as_str().unwrap_or("");
        let bn = b["name"].as_str().unwrap_or("");
        an.to_lowercase().cmp(&bn.to_lowercase())
    });

    Ok(Json(json!({
        "path": canonical.to_string_lossy(),
        "parent": canonical.parent().map(|p| p.to_string_lossy().to_string()),
        "dirs": dirs
    })))
}

File: server-ts/src/routes/fs.ts
import fsNode from 'fs';
import os from 'os';
import path from 'path';
import { execSync } from 'child_process';
import { Router, type Request, type Response } from 'express';

import { ApiError } from '../errors.js';
import { homeDir } from './shared.js';

interface DirEntry {
  name: string;
  path: string;
  isGit: boolean;
}

interface DriveInfo {
  letter: string;
  path: string;
}

/** List available drive letters on Windows (e.g. C:\, D:\). */
function listWindowsDrives(): DriveInfo[] {
  try {
    const raw = execSync(
      'wmic logicaldisk get name',
      { encoding: 'utf-8', stdio: ['pipe', 'pipe', 'pipe'] },
    );
    return raw
      .split(/\r?\n/)
      .map((l) => l.trim())
      .filter((l) => /^[A-Z]:$/.test(l))
      .map((letter) => ({ letter, path: letter + '\\' }));
  } catch {
    return [{ letter: 'C:', path: 'C:\\' }];
  }
}

/** Check if canonical path is a filesystem root. */
function isRootPath(canonical: string): boolean {
  if (process.platform === 'win32') {
    // C:\ or C:
    return /^[A-Za-z]:[/\\]?$/.test(canonical);
  }
  return canonical === '/';
}

export function fsRoutes(): Router {
  const router = Router();

  router.get('/api/fs/list', (req: Request, res: Response) => {
    try {
      const queryPath = (req.query.path as string)?.trim();
      const base = queryPath || homeDir();

      let canonical: string;
      try {
        canonical = fsNode.realpathSync(base);
      } catch {
        return ApiError.badRequest('Cannot resolve path.').toResponse(res);
      }

      const stat = fsNode.statSync(canonical);
      if (!stat.isDirectory()) {
        return ApiError.badRequest('Path is not a directory.').toResponse(res);
      }

      let entries: fsNode.Dirent[];
      try {
        entries = fsNode.readdirSync(canonical, { withFileTypes: true });
      } catch (e) {
        return ApiError.badRequest(
          `Cannot read directory: ${e instanceof Error ? e.message : String(e)}`,
        ).toResponse(res);
      }

      const dirs: DirEntry[] = [];
      for (const entry of entries) {
        if (!entry.isDirectory()) continue;
        if (entry.name.startsWith('.')) continue;
        const fullPath = path.join(canonical, entry.name);
        const isGit = fsNode.existsSync(path.join(fullPath, '.git'));
        dirs.push({ name: entry.name, path: fullPath, isGit });
      }

      dirs.sort((a, b) =>
        a.name.toLowerCase().localeCompare(b.name.toLowerCase()),
      );

      const parentDir = path.dirname(canonical);
      const atRoot = isRootPath(canonical);

      // On Windows at a drive root, include sibling drives for navigation
      const drives =
        process.platform === 'win32' && atRoot
          ? listWindowsDrives()
          : undefined;

      return res.json({
        path: canonical,
        parent: atRoot ? null : parentDir,
        dirs,
        ...(drives ? { drives } : {}),
      });
    } catch (e) {
      if (e instanceof ApiError) return e.toResponse(res);
      return ApiError.internal(e).toResponse(res);
    }
  });

  return router;
}
